02-hello-accessibility/projects/starter/app/src/main/java/com/raywenderlich/android/tacotuesday/discover/DiscoverViewModel.kt

```diff
@@ -0,0 +1,91 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * Notwithstanding the foregoing, you may not use, copy, modify, merge, publish,
+ * distribute, sublicense, create a derivative work, and/or sell copies of the
+ * Software in any work that is designed, intended, or marketed for pedagogical or
+ * instructional purposes related to programming, coding, application development,
+ * or information technology.  Permission for such use, copying, modification,
+ * merger, publication, distribution, sublicensing, creation of derivative works,
+ * or sale is expressly withheld.
+ *
+ * This project and source code may use libraries or frameworks that are
+ * released under various Open-Source licenses. Use of those libraries and
+ * frameworks are governed by their own individual licenses.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.raywenderlich.android.tacotuesday.discover
+
+import androidx.hilt.Assisted
+import androidx.hilt.lifecycle.ViewModelInject
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.raywenderlich.android.tacotuesday.data.Recipe
+import com.raywenderlich.android.tacotuesday.data.RecipeRepository
+import com.raywenderlich.android.tacotuesday.shared.TryDiscardRecipeViewModel
+import com.raywenderlich.android.tacotuesday.shared.TryDiscardRecipeViewModelImpl
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+class DiscoverViewModel @ViewModelInject constructor(
+    private val recipeRepository: RecipeRepository,
+    @Assisted savedStateHandle: SavedStateHandle
+) : ViewModel(), TryDiscardRecipeViewModel by TryDiscardRecipeViewModelImpl(
+    recipeRepository) {
+
+  private val recipe = savedStateHandle.getLiveData<Recipe>(SAVED_STATE_KEY)
+  val nextRecipe = recipe
+
+  private var fetchTacoJob: Job? = null
+
+  fun fetchRandomTaco() {
+    // Reset the timer if this is called again through requesting a new recipe
+    fetchTacoJob?.cancel()
+
+    viewModelScope.launch(Dispatchers.IO) {
+      recipeRepository.randomTacoRecipe()?.let {
+        recipe.postValue(it)
+      }
+    }
+
+    fetchTacoJob = viewModelScope.launch(Dispatchers.IO) {
+      delay(15000) // 15 seconds
```

Comment:
If we're trying to simulate an annoyingly long fetch, I think 5 or 10 seconds is probably better - by 15 almost everyone has bailed in my experience 🙃

Reply:
Ahh maybe I can make this more clear. Rather than a delay, this is the time between auto advancing the recipe to the next one.

The number here is kinda arbitrary, so we can definitely change it. Here's what's going on here.

There are two ways you get a new recipe:
1. You 👍 /👎  it, which will then show a new one
2. After these 15 sec of inactivity, it shows a new one.

We have plans to make it a setting that you can either turn on or off the autoadvance when we're talking about timing. We could think about changing that setting to also include an amount of time. Either way, we can make this shorter.

Reply:
I haven't used coroutines much before, so if you have any suggestions on a better way to do this, I'm happy to hear!

Reply:
@designatednerd @gabrielaa03 I agree with your concerns. This is my first time using coroutines, so I'm struggling to find a better solution. Any suggestions are appreciated! I'm going to comment the code I have better, and add the walkthrough here as well to help conversation.

```kotlin
  fun fetchRandomTaco() {
    // Reset the timer
    // This is so that the timer is reset if the user clicks one of those buttons (which calles this)
    fetchTacoTimer?.cancel()

    // Fetch a random taco
    viewModelScope.launch(Dispatchers.IO) {
      recipeRepository.randomTacoRecipe()?.let {
        recipe.postValue(it)
      }
    }

    // Set up a timer to fetch a new on in 15 seconds
    // This will be reset if the user clicks a try it or discard button
    fetchTacoTimer = viewModelScope.launch(Dispatchers.IO) {
      // Wait 15 seconds on an IO dispatcher (off the main thread)
      delay(15000) // 15 seconds
      // call this function again to fetch a taco and reset the timer!
      fetchRandomTaco()
    }
  }
```

Reply:
This seems reasonable to me, with the caveat that I also am wayyyy behind on coroutines

Reply:
Code seems fine now. I would do a lot of testing just to be sure everything is good.